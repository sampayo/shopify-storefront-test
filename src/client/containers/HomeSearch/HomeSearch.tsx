import React from 'react';
import useTimeout from '@/client/components/hooks/useTimeout';
import { ProductSortKeys } from '@/common/models/product';

export interface IHomeSearchProps {
  className?: string;
  onSubmit?: (query?: { search?: string; sort?: ProductSortKeys }) => void;
}

function HomeSearch(props: IHomeSearchProps) {
  const { onSubmit } = props;
  const { startTimeout, stopTimeout } = useTimeout(400);
  const formDataRef = React.useRef<{ search?: string; sort?: ProductSortKeys }>({});

  const submit = (formData?: { search?: string; sort?: ProductSortKeys }) => {
    stopTimeout();

    if (onSubmit) onSubmit(formData);
  };

  const handleOnSubmit: React.FormEventHandler<HTMLFormElement> = async (e) => {
    if (e?.stopPropagation) e.stopPropagation();
    if (e?.preventDefault) e.preventDefault();

    const formData = new FormData(e.target as any);
    const searchValue = (formData.get('search') as string) || undefined;
    const sortingValue = (formData.get('sorting') as ProductSortKeys) || undefined;

    formDataRef.current = { sort: sortingValue, search: searchValue };
    submit({ ...formDataRef.current });
  };

  const handleOnChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const searchValue = (e.target.value as string) || undefined;

    formDataRef.current = { ...formDataRef.current, search: searchValue };
    startTimeout(() => {
      submit(formDataRef.current);
    });
  };

  const handleOnSortChange: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    const sortValue = (e.target.value as ProductSortKeys) || undefined;

    startTimeout(() => {
      formDataRef.current = { ...formDataRef.current, sort: sortValue };
      submit(formDataRef.current);
    });
  };

  return (
    <form onSubmit={handleOnSubmit} className="gap-4 flex flex-col">
      <input
        type="input"
        placeholder="Search"
        name="search"
        onChange={handleOnChange}
        className="bg-transparent border border-solid border-gray-200 px-4 py-2 rounded-sm w-full"
      ></input>
      <select
        name="sorting"
        defaultValue="RELEVANCE"
        onChange={handleOnSortChange}
        className=" bg-transparent border border-solid border-gray-200 px-4 py-2 rounded-sm max-w-[200PX]"
      >
        <option value="TITLE">Title</option>
        <option value="PRICE">Price</option>
        <option value="RELEVANCE">Relevance</option>
      </select>
      <button type="submit" hidden />
    </form>
  );
}

export default HomeSearch;
