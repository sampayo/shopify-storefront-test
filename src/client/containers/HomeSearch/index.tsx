import HomeSearch from './HomeSearch';

export * from './HomeSearch';

export default HomeSearch;
