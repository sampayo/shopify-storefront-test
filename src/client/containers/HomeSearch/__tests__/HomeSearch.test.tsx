import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import sinon, { SinonSandbox } from 'sinon';
import HomeSearch from '../HomeSearch';

describe('Containers', () => {
  describe('HomeSearch', () => {
    let sandbox: SinonSandbox;
    beforeEach(() => {
      sandbox = sinon.createSandbox();
    });

    afterEach(() => {
      sandbox?.restore();
    });

    test('render home search(snapshot)', () => {
      const tree = renderer.create(<HomeSearch />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('submit without changes', () => {
      const submitSpy = sinon.spy();
      const wrapper = mount(<HomeSearch onSubmit={submitSpy} />);

      const forms = wrapper.find('form');
      expect(forms.length).toEqual(1);
      const form = forms.at(0);

      form.simulate('submit');

      expect(submitSpy.calledOnce).toBeTruthy();
      const call = submitSpy.getCall(0);
      expect(call.args[0]).toEqual({ search: undefined, sort: 'RELEVANCE' });
    });

    test('submit after changing the search values', async () => {
      sandbox.stub(window, 'setTimeout').callsFake((c: any) => c());

      const submitSpy = sandbox.spy();
      const wrapper = shallow(<HomeSearch onSubmit={submitSpy} />);

      const forms = wrapper.find('form');
      expect(forms.length).toEqual(1);
      const searchInputs = wrapper.find('input[type="input"]');
      expect(searchInputs.length).toEqual(1);

      const searchInput = searchInputs.at(0);

      searchInput.simulate('change', { target: { value: 'abcdefg' } });

      expect(submitSpy.calledOnce).toBeTruthy();
      const call = submitSpy.getCall(0);
      expect(call.args[0]).toEqual({ search: 'abcdefg' });
    });
  });
});
