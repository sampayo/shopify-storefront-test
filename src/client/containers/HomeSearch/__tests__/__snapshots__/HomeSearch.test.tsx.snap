// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers HomeSearch render home search(snapshot) 1`] = `
<form
  className="gap-4 flex flex-col"
  onSubmit={[Function]}
>
  <input
    className="bg-transparent border border-solid border-gray-200 px-4 py-2 rounded-sm w-full"
    name="search"
    onChange={[Function]}
    placeholder="Search"
    type="input"
  />
  <select
    className=" bg-transparent border border-solid border-gray-200 px-4 py-2 rounded-sm max-w-[200PX]"
    defaultValue="RELEVANCE"
    name="sorting"
    onChange={[Function]}
  >
    <option
      value="TITLE"
    >
      Title
    </option>
    <option
      value="PRICE"
    >
      Price
    </option>
    <option
      value="RELEVANCE"
    >
      Relevance
    </option>
  </select>
  <button
    hidden={true}
    type="submit"
  />
</form>
`;
